### PR TITLE
feat: use memguard for sensitive data management

### LINKED ISSUE #47: Management of sensitive data in memory
Derived from #11 
>**Lifecycle and protection of secrets:** Ensuring that sensitive cryptographic secrets are zerod/deleted when no longer in use and that they do not persist after they are needed. For instance with GQ signatures, we should delete the RSA signature immediately after the GQ signature is constructed. We should carefully consider the lifecycle of our ephemeral keypairs, currently we just write them to disk.

Potential ways to address this:

1. Avoid using strings and instead use byte slices whenever handling sensitive data
  a. Strings are immutable in Go and difficult to wipe
2. Use the [memguard](https://github.com/awnumar/memguard) package with `LockedBuffers` or `Enclave` objects to handle sensitive data and use the built in data scrubbing methods
  a. This prevents sensitive data from being swapped to disk or leaked through core dumps and ensures it is cleaned up properly
3. Minimize the lifetime of sensitive data in memory by scrubbing immediately after the data is no longer needed

### PR DESCRIPTION
**Pull request by mrjoelkamp**
## Summary
This PR adds [memguard](https://pkg.go.dev/github.com/awnumar/memguard) to manage sensitive data in memory. `memguard` implements a byte slice wipe method that doesn't get optimized out by the compiler and provides a `LockedBuffer` type for locking memory to prevent swapping and handling core dumps.

> * Memory allocation bypasses the language runtime by [using system calls](https://github.com/awnumar/memcall) to query the kernel for resources directly. This avoids interference from the garbage-collector.
> * Buffers that store plaintext data are fortified with guard pages and canary values to detect spurious accesses and overflows.
> * Effort is taken to prevent sensitive data from touching the disk. This includes locking memory to prevent swapping and handling core dumps.
> * Kernel-level immutability is implemented so that attempted modification of protected regions results in an access violation.
> * Multiple endpoints provide session purging and safe termination capabilities as well as signal handling to prevent remnant data being left behind.
> * Side-channel attacks are mitigated against by making sure that the copying and comparison of data is done in constant-time.
> * Accidental memory leaks are mitigated against by harnessing the garbage-collector to automatically destroy containers that have become unreachable.

## Issues

Overall, this feature adds a good amount of protection but was not able to completely cover every instance of sensitive data in memory. There are a few instances in the code based where `memguard`  `LockedBuffers` were not able to be implemented at this time:

## Fixes 
#47

### CODE REVIEW

**Comment by EthanHeilman** [comment]
Excited to see a PR adding memguard! Thanks for putting this together

Really like the evaluation of potential storage of secrets that we can't fully manage with memguard. Can you create an issue for each of the areas you have identified where memguard can not completely control the secrets. That way we can track them and ensure we fix them. I have some ideas on how to address the PK Token issue and I suspect that when we move away from bignum in GQ signatures we might be able to address GQ signatures in bignum  too.

File: client/client.go [hunk 1]
```diff
@@ -57,34 +58,36 @@ func (o *OpkClient) OidcAuth(
...
 
 		sv := gq.NewSignerVerifier(rsaPubKey, GQSecurityParameter)
-		gqToken, err := sv.SignJWT(idToken)
+		gqToken, err := sv.SignJWT(idToken.Bytes())
 		if err != nil {
 			return nil, fmt.Errorf("error creating GQ signature: %w", err)
 		}
 
+		// wipe OIDC signature and replace with GQ signature
+		memguard.WipeBytes(pkt.Op.Signature())
```
[the remark this answers is not in the document]
Comment (mrjoelkamp) [reply client/client.go]: No reason not to defer creation of the PK Token, good find! Refactored in https://github.com/openpubkey/openpubkey/pull/48/commits/ed7d8bfd9e39e8e1e6104538e168ca647425b2e5

**Comment by mrjoelkamp** [comment]
> Really like the evaluation of potential storage of secrets that we can't fully manage with memguard. Can you create an issue for each of the areas you have identified where memguard can not completely control the secrets. That way we can track them and ensure we fix them. I have some ideas on how to address the PK Token issue and I suspect that when we move away from bignum in GQ signatures we might be able to address GQ signatures in bignum too.

Thanks! Created #51 and #52 for the existing issues.

> What are your thoughts on protecting the identity's signing key with memguard?

Created #53 for the identity's signing key since it is similar to #52 using the `math/big` package to store the private key as `big.Int` making `memguard` not viable. It would be good to include it if we figure out a way to make it possible.

The identity's ephemeral signing keys are discussed a bit in #50, prefer using a KMS/HSM for key generation and signing but in the absence of either it would be good to ensure it is handled appropriately in memory.

**Review by EthanHeilman — APPROVED** [review]
LGTM

**Comment by mrjoelkamp** [comment]
I'll need a maintainer to merge this if everything looks good 👍

**Comment by jonnystoten** [comment]
Whoops, I didn't get my comment in before the merge, but I want to push back on this a bit. I think memguard makes it harder to reason about this code without actually protecting us from things like sensitive data being swapped to disk etc.

As noted in the linked issues, if a LockedBuffer's underlying slice is passed to a function that copies it we lose the protection. This will put us in an arms race with every dependency including the standard library - for example, if a patch release of Go changes the implementation of a stdlib function to copy bytes, we'll lose the protection without realizing.

I think it's fine to leave this in as the surface area is pretty limited but I also wanted to share my $0.02

**Pull request merged**